### DISCUSSION TITLE
protobuf definitions in the apk archive

### DISCUSSION THREAD

**Discussion by nlitsme** [tulir/whatsmeow#154] [WhatsApp protocol Q&A]
I found that in the [android .apk for whatsapp](https://apkpure.com/whatsapp-messenger/com.whatsapp) there are all the .proto files, with comments explaining things.

I thought that might give a better result that your [current method](https://github.com/tulir/whatsmeow/tree/main/binary/proto/extract)  of extracting the definitions from the web.whatsapp javascript code

willem

**Comment by nlitsme** [comment]
I noticed it is not in all .apk files,  but in the versions:  2.22.6.7  -  2.22.13.5

[apkcombo](https://apkcombo.com/whatsapp-messenger/com.whatsapp/old-versions/) has is a large list of all old versions

**Reply by nlitsme** [reply]
I have created a git repo with the .proto history extracted from apk files: https://github.com/nlitsme/whatsapp-apk-proto

**Comment by tulir** [comment]
Definitely looks interesting, although it's not exactly compatible with the current schema since it has lots of nesting. Also not sure how well protoc-gen-go deals with those nested types (it would probably create ugly names like `type Message_ImageMessage struct {...}`).
